### PR TITLE
DO NOT MERGE use `input_precision` kwarg for tl.dot 

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128447
* #126098

The `allow_tf32` kwarg has been deprecated. `input_precision` supports a
new mode for f32xf32 matmul utilizing tensor cores without the accuracy
issues normally associated with TF32. Use this mode "tf32x3" if
`torch.backends.cuda.matmul.allow_tf32` is false and it is available.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @desertfire @chauhang

### CODE REVIEW

**Comment by lezcano** [comment]
This needs to be tested for perf and precision, as it's not clear that it provides better precision for small shapes (at least it was not advertised like that when NVIDIA added it to CUTLASS)

**Comment by amjames** [comment]
@lezcano I didn't squash it into the pin update because I wanted to see if other things broke. Updating the triton pin was causing an accuracy issue on one test and using this option seems to have fixed that. 

I had to add [this](https://github.com/pytorch/pytorch/pull/128447/files#diff-8973a6537701d24baabc55ae4c64093f4762a30e4305fddbbbba9d4b54fca3a1R262) condition because mixed-dtype MM kernels were causing triton to segfault during the tritonGPU->llvm conversion.

**Comment by amjames** [comment]
Edited the title, we can't merge this on its own it will have to collapse onto the pin update if things look okay.

File: torch/_inductor/kernel/mm.py
Comment (peterbell10) [inline torch/_inductor/kernel/mm.py:92]: We probably need to be backward compatible with triton versions that don't have `input_precision` yet.

Comment (amjames) [reply torch/_inductor/kernel/mm.py:92]: Do we? I have not seen a policy regarding this, but our published wheels metadata have an exact version pin to the `pytorch-triton` version built from the commit pin at release. 

I think we only need to support the pinned version of triton.

Comment (peterbell10) [reply torch/_inductor/kernel/mm.py:92]: There are several triton pins that aren't all kept in sync. e.g. meta-internal pin and triton-rocm pin

Comment (htyu) [reply torch/_inductor/kernel/mm.py:92]: For internal pins, both the trunk and the Rocm fork is being upgraded to support `input_precision`

File: torch/_inductor/kernel/mm_common.py
Comment (htyu) [inline torch/_inductor/kernel/mm_common.py:265]: Should "tf32x3" be used if allow_tf32=true? I thought when allow_tf32==true, we should use tf32x3 instead of tf32 on the NV path?

**Review by htyu — APPROVED** [review]

**Comment by lezcano** [comment]
@jansel this is what we discussed in San Diego. I think we should test this properly (benchmark accuracy mainly) before merging, but WDYT?

**Comment by amjames** [comment]
closing this as I have folded it into the pin update.

**Comment by jansel** [comment]
This seems fine to me, but we may need some sort of version check on triton to not break BC.

**Pull request closed without merging**
